api/coreservice.go

```diff
@@ -929,28 +929,28 @@ func (core *coreService) readState(ctx context.Context, p protocol.Protocol, hei
 	)
 	ctx = protocol.WithFeatureCtx(protocol.WithFeatureWithHeightCtx(ctx))
 
-	rp := rolldpos.FindProtocol(core.registry)
-	if rp == nil {
-		return nil, uint64(0), errors.New("rolldpos is not registered")
-	}
-
-	tipEpochNum := rp.GetEpochNum(tipHeight)
 	if height != "" {
 		inputHeight, err := strconv.ParseUint(height, 0, 64)
 		if err != nil {
 			return nil, uint64(0), err
 		}
-		inputEpochNum := rp.GetEpochNum(inputHeight)
-		if inputEpochNum < tipEpochNum {
+		rp := rolldpos.FindProtocol(core.registry)
+		if rp != nil {
+			tipEpochNum := rp.GetEpochNum(tipHeight)
+			inputEpochNum := rp.GetEpochNum(inputHeight)
+			if inputEpochNum < tipEpochNum {
+				inputHeight = rp.GetEpochHeight(inputEpochNum)
```

Comment:
can't understand why the height of state is changed in rp, but the changes are  acceptable for me compared to the prev